File: backend/app/schemas/overhaul.py
"""
大修管理数据模式
Overhaul Management Schemas
"""
from pydantic import BaseModel, Field, validator
from typing import Optional, List, Dict, Any
from datetime import date, datetime
from uuid import UUID
from enum import Enum


class OverhaulType(str, Enum):
    """大修类型"""
    SCHEDULED = "scheduled"
    EMERGENCY = "emergency"
    UPGRADE = "upgrade"
    ACCIDENT = "accident"


class OverhaulStatus(str, Enum):
    """大修状态"""
    PLANNING = "planning"
    APPROVED = "approved"
    IN_PROGRESS = "in_progress"
    SUSPENDED = "suspended"
    COMPLETED = "completed"
    CANCELLED = "cancelled"


class OverhaulLevel(str, Enum):
    """大修级别"""
    A1 = "A1"
    A2 = "A2"
    A3 = "A3"
    B1 = "B1"
    B2 = "B2"
    C1 = "C1"
    C2 = "C2"


# ===================== Overhaul Item Schemas =====================
class OverhaulItemBase(BaseModel):
    """大修项目基础模式"""
    item_code: str
    item_name: str
    category: Optional[str] = None
    carriage_number: Optional[str] = None
    component_type: Optional[str] = None
    work_content: Optional[str] = None
    technical_standard: Optional[str] = None
    inspection_method: Optional[str] = None


class OverhaulItemCreate(OverhaulItemBase):
    """创建大修项目"""
    pass


class OverhaulItemUpdate(BaseModel):
    """更新大修项目"""
    status: Optional[str] = None
    progress_percentage: Optional[int] = Field(None, ge=0, le=100)
    quality_check_status: Optional[str] = None
    quality_inspector: Optional[str] = None
    quality_check_date: Optional[date] = None
    quality_notes: Optional[str] = None
    labor_hours: Optional[float] = None
    material_cost: Optional[float] = None
    labor_cost: Optional[float] = None
    total_cost: Optional[float] = None
    spare_parts_used: Optional[List[Dict[str, Any]]] = None
    old_parts_disposal: Optional[str] = None


class OverhaulItemResponse(OverhaulItemBase):
    """大修项目响应"""
    id: UUID
    overhaul_plan_id: UUID
    status: str = "pending"
    progress_percentage: int = 0
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    quality_check_status: Optional[str] = None
    quality_inspector: Optional[str] = None
    quality_check_date: Optional[date] = None
    quality_notes: Optional[str] = None
    labor_hours: Optional[float] = None
    material_cost: Optional[float] = None
    labor_cost: Optional[float] = None
    total_cost: Optional[float] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        orm_mode = True


# ===================== Spare Part Schemas =====================
class SparePartBase(BaseModel):
    """备件基础模式"""
    part_number: str
    part_name: str
    category: Optional[str] = None
    manufacturer: Optional[str] = None
    planned_quantity: int = Field(..., gt=0)
    unit: Optional[str] = None
    unit_price: Optional[float] = None


class SparePartCreate(SparePartBase):
    """创建备件"""
    pass


class SparePartUpdate(BaseModel):
    """更新备件"""
    actual_quantity: Optional[int] = None
    stock_quantity: Optional[int] = None
    warehouse_location: Optional[str] = None
    purchase_order_no: Optional[str] = None
    supplier: Optional[str] = None
    delivery_date: Optional[date] = None
    quality_certificate: Optional[str] = None
    warranty_period: Optional[int] = None


class SparePartResponse(SparePartBase):
    """备件响应"""
    id: UUID
    overhaul_plan_id: UUID
    actual_quantity: Optional[int] = None
    total_price: Optional[float] = None
    stock_quantity: Optional[int] = None
    warehouse_location: Optional[str] = None
    purchase_order_no: Optional[str] = None
    supplier: Optional[str] = None
    delivery_date: Optional[date] = None
    quality_certificate: Optional[str] = None
    warranty_period: Optional[int] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        orm_mode = True


# ===================== Overhaul Plan Schemas =====================
class OverhaulPlanBase(BaseModel):
    """大修计划基础模式"""
    plan_code: str = Field(..., description="大修计划编号")
    train_number: str = Field(..., pattern="^Tr(0[1-9]|1[0-6])$", description="列车编号")
    overhaul_type: OverhaulType
    overhaul_level: OverhaulLevel
    planned_start_date: date
    planned_end_date: date
    estimated_cost: Optional[float] = None
    contractor: Optional[str] = None
    workshop: Optional[str] = None
    responsible_person: Optional[str] = None
    contact_phone: Optional[str] = None
    description: Optional[str] = None
    technical_requirements: Optional[str] = None
    safety_requirements: Optional[str] = None
    quality_standards: Optional[str] = None

    @validator('planned_end_date')
    def validate_dates(cls, v, values):
        if 'planned_start_date' in values and v <= values['planned_start_date']:
            raise ValueError('planned_end_date must be after planned_start_date')
        return v


class OverhaulPlanCreate(OverhaulPlanBase):
    """创建大修计划"""
    vehicle_id: Optional[UUID] = None
    mileage_at_overhaul: Optional[float] = None
    next_overhaul_mileage: Optional[float] = None
    items: Optional[List[OverhaulItemCreate]] = []
    spare_parts: Optional[List[SparePartCreate]] = []


class OverhaulPlanUpdate(BaseModel):
    """更新大修计划"""
    status: Optional[OverhaulStatus] = None
    actual_start_date: Optional[date] = None
    actual_end_date: Optional[date] = None
    actual_cost: Optional[float] = None
    approval_status: Optional[str] = None
    approval_notes: Optional[str] = None
    contractor: Optional[str] = None
    workshop: Optional[str] = None
    responsible_person: Optional[str] = None
    contact_phone: Optional[str] = None


class OverhaulPlanResponse(OverhaulPlanBase):
    """大修计划响应"""
    id: UUID
    vehicle_id: Optional[UUID] = None
    status: OverhaulStatus = OverhaulStatus.PLANNING
    actual_start_date: Optional[date] = None
    actual_end_date: Optional[date] = None
    mileage_at_overhaul: Optional[float] = None
    next_overhaul_mileage: Optional[float] = None
    actual_cost: Optional[float] = None
    currency: str = "CNY"
    approval_status: Optional[str] = None
    approved_by: Optional[UUID] = None
    approved_at: Optional[datetime] = None
    approval_notes: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    created_by: Optional[UUID] = None
    updated_by: Optional[UUID] = None

    # 关联数据
    items: List[OverhaulItemResponse] = []
    spare_parts: List[SparePartResponse] = []

    # 计算字段
    duration_days: Optional[int] = None
    progress_percentage: Optional[float] = None
    cost_variance: Optional[float] = None

    class Config:
        orm_mode = True

    def calculate_fields(self):
        """计算派生字段"""
        # 计算工期
        if self.actual_start_date and self.actual_end_date:
            self.duration_days = (self.actual_end_date - self.actual_start_date).days
        elif self.planned_start_date and self.planned_end_date:
            self.duration_days = (self.planned_end_date - self.planned_start_date).days

        # 计算进度
        if self.items:
            total_progress = sum(item.progress_percentage for item in self.items)
            self.progress_percentage = total_progress / len(self.items)

        # 计算成本偏差
        if self.estimated_cost and self.actual_cost:
            self.cost_variance = self.actual_cost - self.estimated_cost


# ===================== Overhaul Record Schemas =====================
class OverhaulRecordBase(BaseModel):
    """大修记录基础模式"""
    train_number: str
    overhaul_level: OverhaulLevel
    start_date: date
    end_date: date
    mileage_before: Optional[float] = None
    mileage_after: Optional[float] = None
    total_cost: Optional[float] = None
    quality_score: Optional[int] = Field(None, ge=0, le=100)
    performance_improvement: Optional[str] = None
    problems_found: Optional[str] = None
    solutions_applied: Optional[str] = None


class OverhaulRecordCreate(OverhaulRecordBase):
    """创建大修记录"""
    overhaul_plan_id: Optional[UUID] = None
    vehicle_id: Optional[UUID] = None
    labor_cost: Optional[float] = None
    material_cost: Optional[float] = None
    spare_parts_cost: Optional[float] = None
    report_url: Optional[str] = None
    photos: Optional[List[str]] = None
    documents: Optional[List[Dict[str, str]]] = None


class OverhaulRecordResponse(OverhaulRecordBase):
    """大修记录响应"""
    id: UUID
    overhaul_plan_id: Optional[UUID] = None
    vehicle_id: Optional[UUID] = None
    duration_days: Optional[int] = None
    mileage_interval: Optional[float] = None
    labor_cost: Optional[float] = None
    material_cost: Optional[float] = None
    spare_parts_cost: Optional[float] = None
    report_url: Optional[str] = None
    photos: Optional[List[str]] = None
    documents: Optional[List[Dict[str, str]]] = None
    created_at: datetime
    created_by: Optional[UUID] = None

    class Config:
        orm_mode = True


# ===================== Overhaul Standard Schemas =====================
class OverhaulStandardBase(BaseModel):
    """大修标准基础模式"""
    vehicle_type: str
    overhaul_level: OverhaulLevel
    mileage_interval: Optional[float] = None
    time_interval: Optional[int] = None
    whichever_first: bool = True
    standard_duration_days: Optional[int] = None
    standard_cost_min: Optional[float] = None
    standard_cost_max: Optional[float] = None
    technical_requirements: Optional[str] = None
    quality_standards: Optional[str] = None
    acceptance_criteria: Optional[str] = None


class OverhaulStandardCreate(OverhaulStandardBase):
    """创建大修标准"""
    required_items: Optional[List[str]] = None
    optional_items: Optional[List[str]] = None
    applicable_from: Optional[date] = None
    applicable_to: Optional[date] = None


class OverhaulStandardResponse(OverhaulStandardBase):
    """大修标准响应"""
    id: UUID
    required_items: Optional[List[str]] = None
    optional_items: Optional[List[str]] = None
    applicable_from: Optional[date] = None
    applicable_to: Optional[date] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        orm_mode = True


# ===================== Query Schemas =====================
class OverhaulPlanQuery(BaseModel):
    """大修计划查询参数"""
    train_number: Optional[str] = None
    status: Optional[OverhaulStatus] = None
    overhaul_type: Optional[OverhaulType] = None
    overhaul_level: Optional[OverhaulLevel] = None
    contractor: Optional[str] = None
    workshop: Optional[str] = None
    start_date_from: Optional[date] = None
    start_date_to: Optional[date] = None
    page: int = 1
    limit: int = 20


class OverhaulStatistics(BaseModel):
    """大修统计"""
    total_plans: int = 0
    planning: int = 0
    in_progress: int = 0
    completed: int = 0
    total_cost: float = 0
    average_duration: float = 0
    on_time_rate: float = 0
    cost_variance_rate: float = 0
    by_level: Dict[str, int] = {}
    by_type: Dict[str, int] = {}
    upcoming_plans: List[Dict[str, Any]] = []

File: frontend/src/types/overhaul.ts
/**
 * 大修管理类型定义
 * Overhaul Management Type Definitions
 */

// 大修类型
export enum OverhaulType {
  SCHEDULED = 'scheduled',   // 计划大修
  EMERGENCY = 'emergency',   // 紧急大修
  UPGRADE = 'upgrade',       // 升级改造
  ACCIDENT = 'accident'      // 事故维修
}

// 大修状态
export enum OverhaulStatus {
  PLANNING = 'planning',       // 规划中
  APPROVED = 'approved',       // 已批准
  IN_PROGRESS = 'in_progress', // 进行中
  SUSPENDED = 'suspended',     // 暂停
  COMPLETED = 'completed',     // 已完成
  CANCELLED = 'cancelled'      // 已取消
}

// 大修级别
export enum OverhaulLevel {
  A1 = 'A1',  // A1级大修 (架修)
  A2 = 'A2',  // A2级大修 (大修)
  A3 = 'A3',  // A3级大修 (中修)
  B1 = 'B1',  // B1级检修
  B2 = 'B2',  // B2级检修
  C1 = 'C1',  // C1级检修
  C2 = 'C2'   // C2级检修
}

// 大修项目
export interface OverhaulItem {
  id?: string
  overhaul_plan_id?: string
  item_code: string
  item_name: string
  category?: string
  carriage_number?: string
  component_type?: string
  work_content?: string
  technical_standard?: string
  inspection_method?: string
  status?: string
  progress_percentage?: number
  started_at?: string
  completed_at?: string
  quality_check_status?: string
  quality_inspector?: string
  quality_check_date?: string
  quality_notes?: string
  labor_hours?: number
  material_cost?: number
  labor_cost?: number
  total_cost?: number
  spare_parts_used?: any[]
  old_parts_disposal?: string
  created_at?: string
  updated_at?: string
}

// 备件信息
export interface SparePart {
  id?: string
  overhaul_plan_id?: string
  part_number: string
  part_name: string
  category?: string
  manufacturer?: string
  planned_quantity: number
  actual_quantity?: number
  unit?: string
  unit_price?: number
  total_price?: number
  stock_quantity?: number
  warehouse_location?: string
  purchase_order_no?: string
  supplier?: string
  delivery_date?: string
  quality_certificate?: string
  warranty_period?: number
  created_at?: string
  updated_at?: string
}

// 大修计划
export interface OverhaulPlan {
  id?: string
  plan_code: string
  vehicle_id?: string
  train_number: string
  overhaul_type: OverhaulType
  overhaul_level: OverhaulLevel
  status?: OverhaulStatus

  // 时间计划
  planned_start_date: string
  planned_end_date: string
  actual_start_date?: string
  actual_end_date?: string

  // 里程基准
  mileage_at_overhaul?: number
  next_overhaul_mileage?: number

  // 成本预算
  estimated_cost?: number
  actual_cost?: number
  currency?: string

  // 承包信息
  contractor?: string
  workshop?: string
  responsible_person?: string
  contact_phone?: string

  // 审批信息
  approval_status?: string
  approved_by?: string
  approved_at?: string
  approval_notes?: string

  // 其他信息
  description?: string
  technical_requirements?: string
  safety_requirements?: string
  quality_standards?: string

  // 关联数据
  items?: OverhaulItem[]
  spare_parts?: SparePart[]

  // 计算字段
  duration_days?: number
  progress_percentage?: number
  cost_variance?: number

  // 时间戳
  created_at?: string
  updated_at?: string
  created_by?: string
  updated_by?: string
}

// 大修记录
export interface OverhaulRecord {
  id?: string
  overhaul_plan_id?: string
  vehicle_id?: string
  train_number: string
  overhaul_level: OverhaulLevel

  // 执行信息
  start_date: string
  end_date: string
  duration_days?: number

  // 里程信息
  mileage_before?: number
  mileage_after?: number
  mileage_interval?: number

  // 成本信息
  total_cost?: number
  labor_cost?: number
  material_cost?: number
  spare_parts_cost?: number

  // 质量评估
  quality_score?: number
  performance_improvement?: string

  // 问题记录
  problems_found?: string
  solutions_applied?: string

  // 文档附件
  report_url?: string
  photos?: string[]
  documents?: any[]

  created_at?: string
  created_by?: string
}

// 大修标准
export interface OverhaulStandard {
  id?: string
  vehicle_type: string
  overhaul_level: OverhaulLevel

  // 周期标准
  mileage_interval?: number
  time_interval?: number
  whichever_first?: boolean

  // 作业标准
  standard_duration_days?: number
  required_items?: string[]
  optional_items?: string[]

  // 成本标准
  standard_cost_min?: number
  standard_cost_max?: number

  // 技术标准
  technical_requirements?: string
  quality_standards?: string
  acceptance_criteria?: string

  // 适用范围
  applicable_from?: string
  applicable_to?: string

  created_at?: string
  updated_at?: string
}

// 查询参数
export interface OverhaulPlanQuery {
  train_number?: string
  status?: OverhaulStatus
  overhaul_type?: OverhaulType
  overhaul_level?: OverhaulLevel
  contractor?: string
  workshop?: string
  start_date_from?: string
  start_date_to?: string
  page?: number
  limit?: number
}

// 统计信息
export interface OverhaulStatistics {
  total_plans: number
  planning: number
  in_progress: number
  completed: number
  total_cost: number
  average_duration: number
  on_time_rate: number
  cost_variance_rate: number
  by_level: Record<string, number>
  by_type: Record<string, number>
  upcoming_plans: any[]
}

// 工具函数
export const OverhaulUtils = {
  // 获取大修类型标签
  getTypeLabel(type: OverhaulType): string {
    const labels: Record<OverhaulType, string> = {
      [OverhaulType.SCHEDULED]: '计划大修',
      [OverhaulType.EMERGENCY]: '紧急大修',
      [OverhaulType.UPGRADE]: '升级改造',
      [OverhaulType.ACCIDENT]: '事故维修'
    }
    return labels[type] || type
  },

  // 获取大修状态标签
  getStatusLabel(status: OverhaulStatus): string {
    const labels: Record<OverhaulStatus, string> = {
      [OverhaulStatus.PLANNING]: '规划中',
      [OverhaulStatus.APPROVED]: '已批准',
      [OverhaulStatus.IN_PROGRESS]: '进行中',
      [OverhaulStatus.SUSPENDED]: '暂停',
      [OverhaulStatus.COMPLETED]: '已完成',
      [OverhaulStatus.CANCELLED]: '已取消'
    }
    return labels[status] || status
  },

  // 获取大修级别标签
  getLevelLabel(level: OverhaulLevel): string {
    const labels: Record<OverhaulLevel, string> = {
      [OverhaulLevel.A1]: 'A1级大修(架修)',
      [OverhaulLevel.A2]: 'A2级大修',
      [OverhaulLevel.A3]: 'A3级中修',
      [OverhaulLevel.B1]: 'B1级检修',
      [OverhaulLevel.B2]: 'B2级检修',
      [OverhaulLevel.C1]: 'C1级检修',
      [OverhaulLevel.C2]: 'C2级检修'
    }
    return labels[level] || level
  },

  // 获取状态标签类型
  getStatusTagType(status: OverhaulStatus): string {
    const types: Record<OverhaulStatus, string> = {
      [OverhaulStatus.PLANNING]: 'info',
      [OverhaulStatus.APPROVED]: 'warning',
      [OverhaulStatus.IN_PROGRESS]: 'primary',
      [OverhaulStatus.SUSPENDED]: 'danger',
      [OverhaulStatus.COMPLETED]: 'success',
      [OverhaulStatus.CANCELLED]: 'info'
    }
    return types[status] || 'info'
  },

  // 获取级别标签类型
  getLevelTagType(level: OverhaulLevel): string {
    if (level.startsWith('A')) return 'danger'
    if (level.startsWith('B')) return 'warning'
    return 'info'
  },

  // 计算工期
  calculateDuration(startDate: string, endDate: string): number {
    const start = new Date(startDate)
    const end = new Date(endDate)
    const diffTime = Math.abs(end.getTime() - start.getTime())
    return Math.ceil(diffTime / (1000 * 60 * 60 * 24))
  },

  // 计算成本偏差
  calculateCostVariance(estimated: number, actual: number): number {
    if (!estimated || estimated === 0) return 0
    return ((actual - estimated) / estimated) * 100
  },

  // 格式化货币
  formatCurrency(amount: number, currency = 'CNY'): string {
    return new Intl.NumberFormat('zh-CN', {
      style: 'currency',
      currency: currency
    }).format(amount)
  },

  // 获取即将到期的计划
  getUpcomingPlans(plans: OverhaulPlan[], days = 30): OverhaulPlan[] {
    const now = new Date()
    const future = new Date(now.getTime() + days * 24 * 60 * 60 * 1000)

    return plans.filter(plan => {
      if (plan.status !== OverhaulStatus.PLANNING) return false
      const planDate = new Date(plan.planned_start_date)
      return planDate >= now && planDate <= future
    }).sort((a, b) =>
      new Date(a.planned_start_date).getTime() - new Date(b.planned_start_date).getTime()
    )
  }
}